src/control/family.lean

```diff
@@ -0,0 +1,644 @@
+λ i, quot (λ x y, r i (x,y))
+
+namespace quot
+
+variables {α β γ : fam I}  (r : Pred (α ⊗ α))
+
+/-- elimination rule for `fam.quot` -/
+def lift (f : α ⟶ β)
+  (h : ∀ {i} (a : unit i ⟶ α⊗α), a ⊨ r → a ≫ prod.fst ≫ f = a ≫ prod.snd ≫ f) :
+  (quot r ⟶ β) :=
+λ i x, quot.lift (@f i) (λ a b h',
+  let d := value i (fam.subtype r) (subtype.mk (a,b) h') in
+  value.get _ _ (h (value i _ (a,b)) ⟨d,by ext _ ⟨ ⟨ rfl ⟩ ⟩; refl⟩) ) x
+
+/-- introduction rule for `fam.quot` -/
+def mk : α ⟶ quot r :=
+λ (i : I) (x : α i), quot.mk _ x
+
+/-- noncomputable elimination rule for `fam.quot` -/
+noncomputable def out : quot r ⟶ α :=
+λ i x, quot.out x
+
+variables {r}
+
+@[simp, reassoc]
+lemma mk_lift (g : α ⟶ β) (h) :
+  quot.mk r ≫ lift r g h = g :=
+by ext; refl
+
+@[reassoc]
+lemma lift_comp (f : α ⟶ β) (g : β ⟶ γ) (h) :
+  lift r f h ≫ g = lift r (f ≫ g) (by intros; reassoc h; rw h _ a_1) :=
+by { ext, dsimp [lift,(≫)], induction x_1 using quot.ind, refl }
+
+lemma lift_ext (f g : quot r ⟶ β)
+      (hh : quot.mk r ≫ f = quot.mk r ≫ g) :
+  f = g :=
+begin
+  ext a b, apply quot.induction_on b,
+  intros y, apply congr_arrow hh
+end
+
+lemma sound (f : β ⟶ α⊗α)
+      (hh : f ⊨ r) :
+  f ≫ prod.fst ≫ quot.mk r = f ≫ prod.snd ≫ quot.mk r :=
+begin
+  cases hh with f' hh, rw hh,
+  ext, simp [(≫)], apply quot.sound,
+  rcases (f' x_1) with ⟨⟨a,b⟩,h⟩, exact h
+end
+
+lemma sound'' {f g : β ⟶ quot r} (f' g' : β ⟶ α)
+      (hh : diag ≫ fam.prod.map f' g' ⊨ r)
+      (hh_f : f = f' ≫ quot.mk r)
+      (hh_g : g = g' ≫ quot.mk r) :
+  f = g :=
+by { ext; rw [hh_f,hh_g];
+     apply _root_.quot.sound; cases hh with h h',
+     replace h' := congr_arrow h' x_1,
+     simp [fam.prod.map,diag] at h', erw [h'];
+     apply subtype.property }
+
+lemma sound' (f g : β ⟶ α)
+      (hh : diag ≫ fam.prod.map f g ⊨ r) :
+  f ≫ quot.mk r = g ≫ quot.mk r :=
+by apply sound'' f g hh rfl rfl
+
+lemma ind_on (f : β ⟶ quot r) : (∃ g, f = g ≫ quot.mk _) :=
+⟨f ≫ fam.quot.out _, by ext; simp [mk,out]⟩
+
+@[simp, reassoc]
+lemma out_mk (r : Pred (α ⊗ α)) : quot.out r ≫ quot.mk r = 𝟙 _ :=
+by ext; simp [mk,out]; refl
+
+open function
+
+@[simp, reassoc]
+lemma prod.diag_fst : diag ≫ fam.prod.fst = 𝟙 α :=
+by ext; refl
+
+@[simp, reassoc]
+lemma prod.diag_snd : diag ≫ fam.prod.snd = 𝟙 α :=
+by ext; refl
+
+/-- swap the components of a product -/
+def prod.swap : α ⊗ β ⟶ β ⊗ α :=
+diag ≫ (prod.snd ⊗ prod.fst)
+
+@[simp, reassoc]
+lemma prod.swap_fst : prod.swap ≫ fam.prod.fst = (fam.prod.snd : α ⊗ β ⟶ β) :=
+by simp [prod.swap]
+
+@[simp, reassoc]
+lemma prod.swap_snd : prod.swap ≫ fam.prod.snd = (fam.prod.fst : α ⊗ β ⟶ α) :=
+by simp [prod.swap]
+
+/-- reassociate the components of two nested products -/
+def prod.assoc : α ⊗ β ⊗ γ ⟶ α ⊗ (β ⊗ γ) :=
+diag ≫ (prod.fst ≫ prod.fst ⊗ diag ≫ (prod.fst ≫ prod.snd ⊗ prod.snd))
+
+/-- Projection from a product of three components to the
+two left-most components -/
+def lpair : α ⊗ β ⊗ γ ⟶ α ⊗ β := fam.prod.fst
```

Comment:
Yes, it's needed to define `equiv` which in turn lets us define `fam.quot.exact`

Reply:
I mean it's just a special case of another function you've defined. Why not just use `prod.fst`? Same for the others, they are simple compositions and I think it would be better to cut down on all the definitions if possible.

Reply:
I can make them protected to not pollute the global namespace but I think the name I chose clarifies their use in `equiv`.

Reply:
Oh, this was very not clear. You should find a way to make it clearer that `trans` actually says `a = b -> b = c -> a = c` that doesn't involve looking at separate definitions and being a human typechecker.

Reply:
Will do! Maybe the names were not evocative enough. I could try `abc_to_ab`, `abc_to_ac` and `abc_to_bc` if that helps (in addition to a comment)

Reply:
I can reformulate `trans` as:

```lean
(trans : ∀ {i} (abc : unit i ⟶ α ⊗ α ⊗ α), abc ≫ to_ab ⊨ r → abc ≫ to_bc ⊨ r → abc ≫ to_ac ⊨ r)
```